File: scripts/convert_kjv_to_nested.py
#!/usr/bin/env python3
"""
Convert KJV Bible JSON from flat key-value format to nested JSON structure.

Input format: {"Book Chapter:Verse": "Verse text"}
Output format: {"Book": {"Chapter": {"Verse": "Verse text"}}}
"""

import json
import os
import sys
import re

def parse_verse_key(key):
    """Parse a verse key like 'Genesis 1:1' into (book, chapter, verse)."""
    # Use regex to match "Book Chapter:Verse" pattern
    # Handle books with numbers and spaces like "1 Samuel", "Song of Solomon"
    match = re.match(r'^(.+?)\s+(\d+):(\d+)$', key.strip())
    
    if not match:
        return None, None, None
    
    book = match.group(1).strip()
    chapter = match.group(2)
    verse = match.group(3)
    
    return book, chapter, verse

def convert_kjv_to_nested(input_file, output_file):
    """Convert KJV Bible from flat format to nested JSON structure."""
    
    print(f"Reading KJV Bible from: {input_file}")
    
    # Read the input file
    try:
        with open(input_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
    except FileNotFoundError:
        print(f"Error: Could not find input file: {input_file}")
        return False
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in input file: {e}")
        return False
    
    # Initialize the nested structure
    nested_bible = {}
    
    total_verses = len(data)
    print(f"Processing {total_verses} verses...")
    
    skipped_count = 0
    
    for i, (verse_key, verse_text) in enumerate(data.items()):
        # Parse the verse key
        book, chapter, verse = parse_verse_key(verse_key)
        
        if not book or not chapter or not verse:
            print(f"Warning: Could not parse verse key '{verse_key}', skipping")
            skipped_count += 1
            continue
        
        # Initialize nested structure if needed
        if book not in nested_bible:
            nested_bible[book] = {}
        if chapter not in nested_bible[book]:
            nested_bible[book][chapter] = {}
        
        # Add the verse
        nested_bible[book][chapter][verse] = verse_text
        
        # Progress indicator
        if (i + 1) % 5000 == 0:
            print(f"Processed {i + 1}/{total_verses} verses...")
    
    if skipped_count > 0:
        print(f"Warning: Skipped {skipped_count} verses due to parsing issues")
    
    print(f"Conversion complete. Writing to: {output_file}")
    
    # Write the output file
    try:
        with open(output_file, 'w', encoding='utf-8') as f:
            json.dump(nested_bible, f, ensure_ascii=False, indent=2)
    except Exception as e:
        print(f"Error writing output file: {e}")
        return False
    
    # Print summary
    book_count = len(nested_bible)
    verse_count = sum(len(verses) for book in nested_bible.values() for verses in book.values())
    print(f"Successfully converted {book_count} books with {verse_count} verses.")
    
    return True

def main():
    """Main function."""
    # Get the script directory and set up paths
    script_dir = os.path.dirname(os.path.abspath(__file__))
    repo_root = os.path.dirname(script_dir)
    
    # Default paths
    input_file = os.path.join(repo_root, 'bible', 'kjv.json')
    output_file = os.path.join(repo_root, 'bible', 'kjv_nested.json')
    
    # Allow command line arguments for custom paths
    if len(sys.argv) >= 2:
        input_file = sys.argv[1]
    if len(sys.argv) >= 3:
        output_file = sys.argv[2]
    
    print("KJV Bible JSON Converter")
    print("========================")
    print(f"Input:  {input_file}")
    print(f"Output: {output_file}")
    print()
    
    success = convert_kjv_to_nested(input_file, output_file)
    
    if success:
        print("\n✅ Conversion completed successfully!")
        print(f"📖 Nested KJV Bible saved to: {output_file}")
    else:
        print("\n❌ Conversion failed!")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: scripts/convert_kjv_to_nested.js
#!/usr/bin/env node
/**
 * Convert KJV Bible JSON from flat key-value format to nested JSON structure.
 * 
 * Input format: {"Book Chapter:Verse": "Verse text"}
 * Output format: {"Book": {"Chapter": {"Verse": "Verse text"}}}
 */

const fs = require('fs');
const path = require('path');

/**
 * Parse a verse key like 'Genesis 1:1' into {book, chapter, verse}
 */
function parseVerseKey(key) {
  // Use regex to match "Book Chapter:Verse" pattern
  // Handle books with numbers and spaces like "1 Samuel", "Song of Solomon"
  const match = key.trim().match(/^(.+?)\s+(\d+):(\d+)$/);
  
  if (!match) {
    return { book: null, chapter: null, verse: null };
  }
  
  return {
    book: match[1].trim(),
    chapter: match[2],
    verse: match[3]
  };
}

/**
 * Convert KJV Bible from flat format to nested JSON structure
 */
function convertKjvToNested(inputFile, outputFile) {
  console.log(`Reading KJV Bible from: ${inputFile}`);
  
  // Read the input file
  let data;
  try {
    const fileContent = fs.readFileSync(inputFile, 'utf8');
    data = JSON.parse(fileContent);
  } catch (error) {
    if (error.code === 'ENOENT') {
      console.error(`Error: Could not find input file: ${inputFile}`);
      return false;
    } else if (error instanceof SyntaxError) {
      console.error(`Error: Invalid JSON in input file: ${error.message}`);
      return false;
    } else {
      console.error(`Error reading input file: ${error.message}`);
      return false;
    }
  }
  
  // Initialize the nested structure
  const nestedBible = {};
  
  const verseKeys = Object.keys(data);
  const totalVerses = verseKeys.length;
  console.log(`Processing ${totalVerses} verses...`);
  
  let skippedCount = 0;
  
  verseKeys.forEach((verseKey, index) => {
    const verseText = data[verseKey];
    
    // Parse the verse key
    const { book, chapter, verse } = parseVerseKey(verseKey);
    
    if (!book || !chapter || !verse) {
      console.warn(`Warning: Could not parse verse key '${verseKey}', skipping`);
      skippedCount++;
      return;
    }
    
    // Initialize nested structure if needed
    if (!nestedBible[book]) {
      nestedBible[book] = {};
    }
    if (!nestedBible[book][chapter]) {
      nestedBible[book][chapter] = {};
    }
    
    // Add the verse
    nestedBible[book][chapter][verse] = verseText;
    
    // Progress indicator
    if ((index + 1) % 5000 === 0) {
      console.log(`Processed ${index + 1}/${totalVerses} verses...`);
    }
  });
  
  if (skippedCount > 0) {
    console.warn(`Warning: Skipped ${skippedCount} verses due to parsing issues`);
  }
  
  console.log(`Conversion complete. Writing to: ${outputFile}`);
  
  // Write the output file
  try {
    fs.writeFileSync(outputFile, JSON.stringify(nestedBible, null, 2), 'utf8');
  } catch (error) {
    console.error(`Error writing output file: ${error.message}`);
    return false;
  }
  
  // Print summary
  const bookCount = Object.keys(nestedBible).length;
  const verseCount = Object.values(nestedBible)
    .reduce((total, book) => 
      total + Object.values(book).reduce((bookTotal, chapter) => 
        bookTotal + Object.keys(chapter).length, 0), 0);
  
  console.log(`Successfully converted ${bookCount} books with ${verseCount} verses.`);
  
  return true;
}

function main() {
  // Get the script directory and set up paths
  const scriptDir = __dirname;
  const repoRoot = path.dirname(scriptDir);
  
  // Default paths
  let inputFile = path.join(repoRoot, 'bible', 'kjv.json');
  let outputFile = path.join(repoRoot, 'bible', 'kjv_nested.json');
  
  // Allow command line arguments for custom paths
  const args = process.argv.slice(2);
  if (args.length >= 1) {
    inputFile = args[0];
  }
  if (args.length >= 2) {
    outputFile = args[1];
  }
  
  console.log('KJV Bible JSON Converter (Node.js)');
  console.log('==================================');
  console.log(`Input:  ${inputFile}`);
  console.log(`Output: ${outputFile}`);
  console.log();
  
  const success = convertKjvToNested(inputFile, outputFile);
  
  if (success) {
    console.log('\n✅ Conversion completed successfully!');
    console.log(`📖 Nested KJV Bible saved to: ${outputFile}`);
  } else {
    console.log('\n❌ Conversion failed!');
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

module.exports = { convertKjvToNested, parseVerseKey };